### PR TITLE
adapter: Fix timeline determination for temporals

### PR DESCRIPTION
**Pull request by jkosh44**
Previously we would determine the timeline context of a query by performing BFS over all the objects in the query and their dependencies. At each dependency we would add a timeline using the following logic:

  - If the current object is a leaf view, and we HAVE previously seen a view with a temporal function then add the `TimestampDependent` timeline context.
  - If the current object is a leaf view, and we HAVE NOT previously seen a view with a temporal function then add the `TimestampIndependent` timeline context.
  - All other objects directly add their timeline.

The final timeline contexts from this process is dependent on the traversal order of each node. Specifically if a view with a temporal function is visited after all leaf views have been visited, then we will incorrectly not add any `TimestampDependent` contexts.

This commit fixes this issue by changing the logic to the following:

  - If the current view contains a temporal function then add the `TimestampDependent` timeline context.
  - If the current view does not contain a temporal function then add the `TimestampIndependent` timeline context.
  - All other objects directly add their timeline.

### Motivation
This PR fixes a previously unreported bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Fix query results that rely on static views with temporal filters.

### CODE REVIEW

**Review by def- — CHANGES_REQUESTED** [review]
Please also add the test scenario that @teskje posted as a slt to make sure this doesn't regress

File: src/environmentd/tests/sql.rs [hunk 1]
```diff
@@ -3645,3 +3645,102 @@ async fn test_retain_history() {
         );
     }
 }
+
+#[mz_ore::test(tokio::test(flavor = "multi_thread", worker_threads = 1))]
+async fn test_temporal_static_queries() {
```
Comment (def-) [inline src/environmentd/tests/sql.rs:3650]: Hm, does this have an advantage over using slt directly?

Comment (jkosh44) [reply src/environmentd/tests/sql.rs:3650]: The problem with creating an SLT test, is that the answer changes depending on the current time. I couldn't easily figure out a way to write an SLT test that wouldn't break in the future. Any ideas?

Comment (ggevay) [reply src/environmentd/tests/sql.rs:3650]: Maybe we could change Jan's repro to make `mz_now()` comparisons only to dates very far in the future, e.g., in 2040 or something.

Comment (def-) [reply src/environmentd/tests/sql.rs:3650]: Right, maybe testdrive with `$ set-regex` would help but that's fine then.

**Review by def- — APPROVED** [review]

File: test/sqllogictest/temporal.slt
Comment (jkosh44) [inline test/sqllogictest/temporal.slt:1]: @def- This should work for the next 8,000 years hopefully.

**Review by ggevay — APPROVED** [review]
Thanks!

**Comment by umanwizard** [comment]
Looks fine, but why not get rid of the internal function?

```
diff --git a/src/adapter/src/coord/timeline.rs b/src/adapter/src/coord/timeline.rs
index c47870d0c7..4d23da2494 100644
--- a/src/adapter/src/coord/timeline.rs
+++ b/src/adapter/src/coord/timeline.rs
@@ -17,7 +17,7 @@ use std::sync::Arc;
 use chrono::{DateTime, Utc};
 use itertools::Itertools;
 use mz_adapter_types::connection::ConnectionId;
-use mz_catalog::memory::objects::CatalogItem;
+use mz_catalog::memory::objects::{CatalogItem, View, MaterializedView};
 use mz_compute_types::ComputeInstanceId;
 use mz_expr::{CollectionPlan, OptimizedMirRelationExpr};
 use mz_ore::collections::CollectionExt;
@@ -496,20 +496,6 @@ impl Coordinator {
         // the set until we reach the bottom.
         let mut ids: Vec<_> = ids.into_iter().collect();
         // Helper function for both materialized views and views.
-        fn visit_view_expr(
-            optimized_expr: &OptimizedMirRelationExpr,
-            ids: &mut Vec<GlobalId>,
-            timelines: &mut BTreeSet<TimelineContext>,
-        ) {
-            // If the definition contains a temporal function, the timeline must
-            // be timestamp dependent.
-            if optimized_expr.contains_temporal() {
-                timelines.insert(TimelineContext::TimestampDependent);
-            } else {
-                timelines.insert(TimelineContext::TimestampIndependent);
-            }
-            ids.extend(optimized_expr.depends_on());
-        }
         while let Some(id) = ids.pop() {
             // Protect against possible infinite recursion. Not sure if it's possible, but
             // a cheap prevention for the future.
@@ -525,11 +511,16 @@ impl Coordinator {
                     CatalogItem::Index(index) => {
                         ids.push(index.on);
                     }
-                    CatalogItem::View(view) => {
-                        visit_view_expr(&view.optimized_expr, &mut ids, &mut timelines);
-                    }
-                    CatalogItem::MaterializedView(mview) => {
-                        visit_view_expr(&mview.optimized_expr, &mut ids, &mut timelines);
+                    CatalogItem::View(View { optimized_expr, .. })
+                    | CatalogItem::MaterializedView(MaterializedView { optimized_expr, .. }) => {
+                        // If the definition contains a temporal function, the timeline must
+                        // be timestamp dependent.
+                        if optimized_expr.contains_temporal() {
+                            timelines.insert(TimelineContext::TimestampDependent);
+                        } else {
+                            timelines.insert(TimelineContext::TimestampIndependent);
+                        }
+                        ids.extend(optimized_expr.depends_on());
                     }
                     CatalogItem::Table(table) => {
                         timelines.insert(TimelineContext::TimelineDependent(table.timeline()));
```

**Comment by teskje** [comment]
Thanks for fixing!

I have a question though: IIUC the new version of this function outputs just a set of timelines and the only guarantee is that if there are temporal filters in the input objects, at least one of them will be `TimestampDependent` (but for some objects timelines might vary between dependent/independent depending on in which order the graph is visited). Does that provide any value over just having a `is_timestamp_dependent(ids) -> bool` that immediately returns once it has found a single temporal filter?

**Comment by jkosh44** [comment]
> Looks fine, but why not get rid of the internal function?
> 
> ```
> diff --git a/src/adapter/src/coord/timeline.rs b/src/adapter/src/coord/timeline.rs
> index c47870d0c7..4d23da2494 100644
> --- a/src/adapter/src/coord/timeline.rs
> +++ b/src/adapter/src/coord/timeline.rs
> @@ -17,7 +17,7 @@ use std::sync::Arc;
>  use chrono::{DateTime, Utc};
>  use itertools::Itertools;
>  use mz_adapter_types::connection::ConnectionId;
> -use mz_catalog::memory::objects::CatalogItem;
> +use mz_catalog::memory::objects::{CatalogItem, View, MaterializedView};
>  use mz_compute_types::ComputeInstanceId;
>  use mz_expr::{CollectionPlan, OptimizedMirRelationExpr};
>  use mz_ore::collections::CollectionExt;
> @@ -496,20 +496,6 @@ impl Coordinator {
>          // the set until we reach the bottom.
>          let mut ids: Vec<_> = ids.into_iter().collect();
>          // Helper function for both materialized views and views.
> -        fn visit_view_expr(
> -            optimized_expr: &OptimizedMirRelationExpr,
> -            ids: &mut Vec<GlobalId>,
> -            timelines: &mut BTreeSet<TimelineContext>,
> -        ) {
> -            // If the definition contains a temporal function, the timeline must
> -            // be timestamp dependent.
> -            if optimized_expr.contains_temporal() {
> -                timelines.insert(TimelineContext::TimestampDependent);
> -            } else {
> -                timelines.insert(TimelineContext::TimestampIndependent);
> -            }
> -            ids.extend(optimized_expr.depends_on());
> -        }
>          while let Some(id) = ids.pop() {
>              // Protect against possible infinite recursion. Not sure if it's possible, but
>              // a cheap prevention for the future.
> @@ -525,11 +511,16 @@ impl Coordinator {
>                      CatalogItem::Index(index) => {
>                          ids.push(index.on);
>                      }
> -                    CatalogItem::View(view) => {
> -                        visit_view_expr(&view.optimized_expr, &mut ids, &mut timelines);
> -                    }
> -                    CatalogItem::MaterializedView(mview) => {
> -                        visit_view_expr(&mview.optimized_expr, &mut ids, &mut timelines);
> +                    CatalogItem::View(View { optimized_expr, .. })
> +                    | CatalogItem::MaterializedView(MaterializedView { optimized_expr, .. }) => {
> +                        // If the definition contains a temporal function, the timeline must
> +                        // be timestamp dependent.
> +                        if optimized_expr.contains_temporal() {
> +                            timelines.insert(TimelineContext::TimestampDependent);
> +                        } else {
> +                            timelines.insert(TimelineContext::TimestampIndependent);
> +                        }
> +                        ids.extend(optimized_expr.depends_on());
>                      }
>                      CatalogItem::Table(table) => {
>                          timelines.insert(TimelineContext::TimelineDependent(table.timeline()));
> ```

Good point, done.

**Review by umanwizard — APPROVED** [review]

**Comment by jkosh44** [comment]
> I have a question though: IIUC the new version of this function outputs just a set of timelines

FWIW, the old version also outputed just a set of timelines.

> and the only guarantee is that if there are temporal filters in the input objects, at least one of them will be TimestampDependent

Also FWIW, this was the intended behavior of the old version.

> (but for some objects timelines might vary between dependent/independent depending on in which order the graph is visited).

I don't think this is true if I understand correctly. We visit every node exactly once, the logic we apply at each node is now idempotent, and we return a `BTreeSet` so the order of the result will be consistent. So no matter what order the graph is visited, we'll always return the same set of timelines.

> Does that provide any value over just having a is_timestamp_dependent(ids) -> bool that immediately returns once it has found a single temporal filter?

There are three `TimelineContext` variants, `TimelineDependent`, `TimestampDependent`, `TimestampIndependent`. What we would actually need is something like

```Rust
let timeline_context =  if let Some(timeline) = is_timeline_dependent(ids) {
  TimelineContext::TimelineContext(timeline)
} else if is_timestamp_dependent(ids) {
  TimelineContext::TimestampDependent
} else {
  TimelineContext::TimestampIndependent
}
```

So in the worst case, we end up going through every ID twice. Additionally, the function that is updated in the PR is also used to validate that a query does not mix two different timelines. So it already has to go through every node.

**Comment by teskje** [comment]
> I don't think this is true if I understand correctly. We visit every node exactly once, the logic we apply at each node is now idempotent, and we return a BTreeSet so the order of the result will be consistent. So no matter what order the graph is visited, we'll always return the same set of timelines.

You are right, I missed that we got rid of `contains_temporal` now.

All checks out then!

**Pull request merged**